src/screencast/pipewire_screencast.c

```diff
@@ -8,27 +9,68 @@
 #include <sys/mman.h>
 #include <unistd.h>
 #include <assert.h>
+#include <libdrm/drm_fourcc.h>
+#include "linux-dmabuf-unstable-v1-client-protocol.h"
 
 #include "wlr_screencast.h"
 #include "xdpw.h"
 #include "logger.h"
 
-static struct spa_pod *build_format(struct spa_pod_builder *b, enum spa_video_format format,
-		uint32_t width, uint32_t height, uint32_t framerate) {
+static struct spa_pod *build_buffer(struct spa_pod_builder *b, uint32_t blocks, uint32_t size,
+		uint32_t stride, uint32_t datatype) {
+	assert(blocks > 0);
+	assert(datatype > 0);
 	struct spa_pod_frame f[1];
 
+	spa_pod_builder_push_object(b, &f[0], SPA_TYPE_OBJECT_ParamBuffers, SPA_PARAM_Buffers);
+	spa_pod_builder_add(b, SPA_PARAM_BUFFERS_buffers, SPA_POD_CHOICE_RANGE_Int(XDPW_PWR_BUFFERS, 1, 32), 0);
+	spa_pod_builder_add(b, SPA_PARAM_BUFFERS_blocks, SPA_POD_Int(blocks), 0);
+	if (size > 0)
+		spa_pod_builder_add(b, SPA_PARAM_BUFFERS_size, SPA_POD_Int(size), 0);
+	if (stride > 0)
+		spa_pod_builder_add(b, SPA_PARAM_BUFFERS_stride, SPA_POD_Int(stride), 0);
+	spa_pod_builder_add(b, SPA_PARAM_BUFFERS_align, SPA_POD_Int(XDPW_PWR_ALIGN), 0);
+	spa_pod_builder_add(b, SPA_PARAM_BUFFERS_dataType, SPA_POD_CHOICE_FLAGS_Int(datatype), 0);
+	return spa_pod_builder_pop(b, &f[0]);
+}
+
+static struct spa_pod *build_format(struct spa_pod_builder *b, enum spa_video_format format,
+		uint32_t width, uint32_t height, uint32_t framerate,
+		uint64_t *modifiers, int modifier_count) {
+	struct spa_pod_frame f[2];
+	int i, c;
+
 	enum spa_video_format format_without_alpha = xdpw_format_pw_strip_alpha(format);
 
 	spa_pod_builder_push_object(b, &f[0], SPA_TYPE_OBJECT_Format, SPA_PARAM_EnumFormat);
 	spa_pod_builder_add(b, SPA_FORMAT_mediaType, SPA_POD_Id(SPA_MEDIA_TYPE_video), 0);
 	spa_pod_builder_add(b, SPA_FORMAT_mediaSubtype, SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw), 0);
 	/* format */
-	if (format_without_alpha == SPA_VIDEO_FORMAT_UNKNOWN) {
+	if (modifier_count > 0 || format_without_alpha == SPA_VIDEO_FORMAT_UNKNOWN) {
+		// modifiers are defined only in combinations with their format
+		// we should not announce the format without alpha
 		spa_pod_builder_add(b, SPA_FORMAT_VIDEO_format, SPA_POD_Id(format), 0);
 	} else {
 		spa_pod_builder_add(b, SPA_FORMAT_VIDEO_format,
 				SPA_POD_CHOICE_ENUM_Id(3, format, format, format_without_alpha), 0);
 	}
+	/* modifiers */
+	if (modifier_count == 1 && modifiers[0] == DRM_FORMAT_MOD_INVALID) {
+		// we only support implicit modifiers, use shortpath to skip fixation phase
+		spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_modifier, SPA_POD_PROP_FLAG_MANDATORY);
+		spa_pod_builder_long(b, modifiers[0]);
```

Comment:
I think we need to duplicate this line. The first one is the default, the rest is the full list of possible choices.

Reply:
This is the "fast" path so only one modifier and no SPA_CHOICE_Enum. Analog to announcing the supported format. I can compactify it to 
```
spa_pod_builder_add(b, SPA_FORMAT_VIDEO_modifier, SPA_POD_Long(modifiers[0]), SPA_POD_PROP_FLAG_MANDATORY);
```
.

Reply:
Oh right, I see, that makes sense! Current code is fine by me.